### DISCUSSION TITLE
Save and restore tab configurations

### DISCUSSION THREAD

**Discussion by localauthor** [alphapapa/activities.el#78] [Ideas]
I am wondering if activities.el could be used to save and restore tab configurations (similar to Safari's "Tab Groups," which allows you to name and restore sets of browser tabs for future sessions).

Perhaps I'm overlooking something obvious, but I cannot find any existing functionality that handles tab configurations by frame. There are a lot of ways to handle window configurations, but what about tabs?
Any thoughts or leads? Thanks.

**Comment by shipmints** [comment]
Adam @alphapapa, thank you for this package and your others. Having a more modern and streamlined workspace-management package in the spirit of @mindad's individual packages makes a lot of sense. Kudos to you for not bloating ```activities``` with features that can, and should, be in other packages.

Kindling/rekindling this discussion. Also per https://www.reddit.com/r/emacs/comments/1bs9bri/comment/kxjkh5l/

I (and several colleagues) tend to use Emacs with a project per frame with multiple tabs with window configurations in each frame. Think about one tab for editing code, a second arranged to run code and perhaps show charts, a third arranged to debug, a fourth arranged to interact with a "forge" and perhaps "developer-oriented social media (aka project manager fodder or friendly discussions like this one)."

What I've been having to do is set up a frame with four ```activities```, each with its own name, and then hand-craft their restoration to pick up where I left off. Create empty frame. Restore activity 1. Restore 2. Restore 3. Restore 4. I agree it is nice not to be forced to restore all of them if I need only, say, the forge.

If I could treat them both individually and also collect them in a group for higher-level management, perhaps called a "meta activity" or "activity group" that would be ideal.

I occasionally will also launch a fresh frame (sometimes on another display) to load one of the activities and if it is already in use, I need to "disconnect" it from the existing frame, rather than "reparent" it in a new or existing frame which would be helpful but less urgent than activity groups as a stored meta activity.

@localauthor I am curious if the use-case outlined here is similar to the one you envision.

I would love to contribute this code but I'm more of a hack Emacs into submission type than a hard-core elisper. I could try with some guidance, perhaps, once an approach is agreed. I do see @jdtsmith hacking on ```activities``` which is great. He's another great contributor to the community.

Regards, and looking forward to using ```activities``` for a long time to come as it seems destined to be a modern classic @shipmints

This seems related to discussions in #3 and also #84, perhaps others. Glad I'm not alone!

**Reply by alphapapa** [reply]
@shipmints Thanks for the kind words.  I'm very glad to hear that Activities is proving useful.

I'm fully in favor of adding functionality something like what you've described.  Please see https://github.com/alphapapa/activities.el/issues/84#issuecomment-2177803750.  I'd be glad for someone to help refine the ideas before implementation is started.

**Comment by shipmints** [comment]
One of the ```activities``` features that confused me is that it names the current frame with most-recent activity defined in that frame. The second activity, on a separate tab, does not change the frame name. That led me to initially think that ```activities``` was frame-wide. Perhaps this is an unintended feature and while multi-tab and multi-frame activities are not supported, the frame title should change when the tab is switched.

**Reply by jdtsmith** [reply]
`activities-set-frame-name` is only supposed to apply when not using tabs for frames.

**Reply by shipmints** [reply]
Seems the policy in ```activities-switch``` is set it always when ```activities-set-frame-name``` is non-nil. I will disable the flag ```activities-set-frame-name``` until these policies are more ironed out. Thanks for taking a look.

**Reply by alphapapa** [reply]
@shipmints If you think there's a bug or unexpected behavior here, please file an issue about it specifically.  I only use `activities-tabs-mode` myself, so I'm less likely to notice issues with frames.

**Comment by jdtsmith** [comment]
A simple :

```elisp
(do-list (a '("my" "favorite" "activities")) (activities-resume (activities-named a)))
```

should work, perhaps bound to a key.    This rapid resume loop uncovered a bug for me; see #98.  But note that I use frames exclusively (mapped to native tabs in emacs-mac), so YMMV with tabs. 

For moving an activity tab to another frame, can you not just use something like `tab-bar-move-tab-to-frame`?

**Reply by shipmints** [reply]
Perhaps a non-blocking delay (maybe sit-for? or the delay equivalent run-with-idle timer for each element) inside the loop might help avoid the race condition you found?

Wrt ```tab-bar-move-tab-to-frame``` it's unfortunate that it uses frame number rather than frame name and doesn't prompt to select a frame. But, yes, it is there and it does work. Thanks for that tip.

**Reply by jdtsmith** [reply]
It's fixed in #98.

**Reply by shipmints** [reply]
Thank you.

**Comment by localauthor** [comment]
@shipmints Your use case sounds quite a bit like mine: one "project" per frame, with pertinent files opened in different tab-bar tabs.

For example, when preparing a lecture, I have one tab with lecture notes, one tab with a slide show config, and one with general notes. Because I teach the same class multiple times, I manually restore this configuration the next time I teach. It's the "manually restore" step that I'm inclined to eliminate.

I have not incorporated activities.el into this process, so maybe I'm missing out on something useful.

But, I have rolled a rudimentary solution for myself that I've dubbed [tab-sets](https://github.com/localauthor/tab-sets). It uses the built-in `frameset` and `tab-bar` libraries, along with some elbow grease, and does exactly (and only) what I want, so far. That is, I can save sets of tabs (giving each tab-set a distinct name) and then reopen them at a later time. (Tab-sets are saved to a data file, so they persist across emacs sessions.)

I don't have plans to publish it as a package, but maybe some of its widgetry could be incorporated into =activities.el= (by you or someone else), to get you where you want to be?

**Reply by shipmints** [reply]
Not that I'm cheating on ```activities```...I've been experimenting with @overideal's https://github.com/overideal/perject https://www.reddit.com/r/emacs/comments/11pcvy9/perject_is_now_available_on_melpa/

If you can get over the concept of "project" being overloaded, it seems pretty feature complete and uses only native Emacs facilities and ```transient``` (which pretty much everyone has installed anyway). The default key bindings suggest the author uses Windows or at least uses has a full-sized keyboard with a insert/delete/home/end cluster, so those probably need changing in your own set up.

There have been several issues and I'm trying to work through them. Buffers aren't automatically added to an open "project" when they should be, and they are sometimes added which confuses me. The behavior of a "collection" without a "project" seems wonky. I can't make heads or tails out what ```perject-tab``` is really for. When ```tab-bar``` is properly configured (i.e., history enabled) it works well. ```perject-tab``` doesn't seem to know about ```tab-bar``` tab groups which I have automatically set to the native ```project-current``` of the current buffer. I use ```tab-bar``` commands like C-x t 2 and C-x t 0 all the time and not clear if I have to use the ```perject-tab``` versions of those commands or not and how it would know if I'm in a collection or not when using those commands. ```perject``` use of ```desktop-save``` doesn't seem to restore frame parameters correctly. I use maximized (not horrid "full screen") frames all of the time and they are positioned upper left all the time. Restored frames show up many pixels below and to the right of where they originated. Perhaps ```desktop``` itself is busted. I'd happily screw with ```frameset-filter-alist``` if that's helpful and see if ```default-frame-alist``` takes precedence. It is not clear if a collection can span frames. Or if projects are frame specific. Or the other limitations/design intents. Digging through the code to see. I have not yet been able to get a collection restored with all the stuff that was supposed to be in when I last saved it. I will plug away.

Once I understand more, I will open a discussion with @overideal to see if perhaps there have been changes/fixes that have been made but not yet published and to understand his own work flow. Does he ever use a collection without a project? If not, why does it exist if it's impractical and what is the behavior supposed to be.

Anyway, worth taking a gander by all of us. The guy put a lot of work into it. I'm happy to contribute there, also.

P.S. ```init-lock``` might have saved me years of my life. Tweaking my Emacs configuration is now a permanent part of my character.

**Reply by overideal** [reply]
@shipmints thanks for the kind words. If you are still interested, feel free to open a discussion in the perject repo and I will answer any questions.

**Comment by alphapapa** [comment]
FWIW, the status quo of this issue with regard to `activities` is here: https://github.com/alphapapa/activities.el/discussions/106#discussioncomment-9848930  Please consolidate further discussion there.

I'm closing this discussion (not to devalue anything said here, just to consolidate, because right now it's happening in at least 3 places).
